### DISCUSSION TITLE
Chained Tasks Not Executed In-order

### DISCUSSION THREAD

**Discussion by arshad171** [celery/celery#7666] [Q&A] [answered]
### Discussed in https://github.com/celery/celery/discussions/7114
<div type='discussions-op-text'>

<sup>Originally posted by **arshad171** November 22, 2021</sup>
For some of my TensorFlow pipelines, I need the task chains to be executed **in the order they were received**, i.e. execute the entire chain before beginning with the next chain scheduled.

I have abstracted the problem with the example below,

## Sample Tasks
```
@app.task
def task1(id):
    print(f'{id=} | task 1')
    time.sleep(5)
    return id

@app.task
def task2(id):
    print(f'{id=} | task 2')
    time.sleep(5)
    return id
```

## Chaining Tasks
Creating two chains, with different ids, `id=1` and `id=2`.

```
res1 = chain(
    task1.s(1),
    task2.s()
).apply_async()

res2 = chain(
    task1.s(2),
    task2.s()
).apply_async()
```

### The result I expect: The first chain (with id=1) to be processed before the chain with id=2 is picked up.
```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### What I observe: task1 from both the chains are run first, and then the task2 is picked up.
```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### I have also tried passing `--concurrency=1` and `--prefetch-multiplier=1` when starting the worker, but it doesn't seem to help, same result as above.
Is this the expected behaviour?

## Setup
* Python: 3.8
* Celery: 5.2.1 (latest pypi release)
* Backend: RabbitMQ
* OS: Linux (Manjaro)

Thanks</div>

[a comment by an automated account is not shown]

**Comment by arshad171** [comment]
any update?

**Comment by glivne** [comment]
You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?

To clarify - if you're looking for these tasks to run all in order, a single chain sounds like a better solution.

**Comment by arshad171** [comment]
> You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?
> 
> To clarify - if you're looking for these tasks to run all in order, a single chain sounds like a better solution.

can you please suggest the celery API to do that?

my assumption was that `chain(task1, task2)` would link the tasks, and calling `apply_async` would execute that task group asynchronously.
https://docs.celeryproject.org/en/stable/userguide/canvas.html#chains

**Comment by glivne** [comment, marked as answer]
> > You're calling apply_async. why would you assume all tasks of chain 1 would finish before chain 2?
> > To clarify - if you're looking for these tasks to run all in order, a single chain sounds like a better solution.
> 
> can you please suggest the celery API to do that?
> 
> my assumption was that `chain(task1, task2)` would link the tasks, and calling `apply_async` would execute that task group asynchronously. https://docs.celeryproject.org/en/stable/userguide/canvas.html#chains

That is correct. But there is no guarantee between the relative order of tasks outside of each of the chains.
So task1,1 will indeed always execute before task1,2 . And the output tuple args of task 1 will be the first arguments to task 2.
So what you could do, is 1 big chain, if it's okay that task 1,2 will pass its output to task 2,1. Like:

```
chain_result = chain([
 task1.s(1),
 task2.s(),
 task1.s(id=2),
 task2.s()
]).apply_async()
```

If you want to keep the logic separated, you could do a chain of 2 chains. as such:

```
full_chain_result = chain([
 chain(task1.s(1), task2),
 chain(task1.s(2), task2)
]).apply_async()
```
alternatively, you could just wait for chain 1 to finish before calling chain2:
```
res1 = chain(
    task1.s(1),
    task2.s()
).apply_async()
res1.get()
res2 = chain(
    task1.s(2),
    task2.s()
).apply_async()
```

Let me know if any of those worked.

**Comment by arshad171** [comment]
@glivne thank you for the quick response.

The use-case I'm working with involves the client sending independent requests (via REST requests) that need to be processed asynchronously. 
And I have provided a toy example modeling the problem and the result I expect.
- the client sends request to execute a set of tasks (task1 and task2).
- the client expects a async response, so the execution has to be non-blocking.

**solution 1 and 2**: the next request to execute the task chain (task1 and task2) will not be known prior to it's arrival, it can an independent request. hence this will not be possible

solution 3: the tasks I'm performing are very time consuming tasks, and it would better to perform them async. Again the subsequest request to perform the chain (task1, task2) will not be known prior to the client request.

**Comment by arshad171** [comment]
the solution to my problem is simple, "First In First Out" Scheduling, but I'm not sure if celery exposes a way to do this.

**Reply by anthonytwh** [reply]
@arshad171 did you find a solution to this? I'm having the same issue when a chain of tasks are being executed by an API. New requests start before a running chain is complete.

I have tried limiting the celery configuration to 1 queue, but the chains do not seem to be appended to the queue in FIFO. Adding priorities to the tasks did not work either.

**Reply by norbertcyran** [reply]
@anthonytwh In this case, do you really need to split your task into a chain? Chains should ensure order within the chain itself, but do not guarantee order between separate chains. In your situation, maybe one longer task would be enough?

**Reply by anthonytwh** [reply]
@norbertcyran that might be possible. The chain I have has multiple long-running tasks that have varying times to complete, and a final task that waits for those tasks to finish to execute. The long running tasks are in a group in the main chain, and it might be possible to sequence it all as a single tasks but it wouldn't be efficient. It is also helpful to monitor Flower for these long-running tasks, if one were to have issues it is easier to identify than if it were in a giant single task.

Is there a setting to force Celery to wait for existing chains to complete? (ie. queuing chains by order of request time)

**Reply by norbertcyran** [reply]
So there is a group inside, I see.

> Is there a setting to force Celery to wait for existing chains to complete?

No, I don't think so.

Could you briefly describe why those chains can't be executed in parallel? Do they work on the same data, is it some periodic processing or so? Depending on the use case, there might be different solutions to the problem.

The first thing that comes to my mind is a distributed lock, e.g. with redis or Memcached: https://redis-py.readthedocs.io/en/v4.5.1/lock.html. You could then acquire a lock at the beginning of the chain, and release it at the end. Though be careful about possible deadlocks: acquiring task should not wait until the lock is available, but rather retry after some time if the lock is already acquired. This could be accomplished with `blocking=False` argument to redis `Lock`. Also, make sure to set an appropriate timeout.

**Reply by anthonytwh** [reply]
@norbertcyran The long tasks that are run as a group are executing commands to specified containers (runners) to execute those commands, and these runners rely on local caching of data and results (files). When the chains are running in parallel, the runners are receiving the commands as expected but the local caching of data and results are getting mixed up. There are also tasks at the end of the chain that will perform a cleanup of local cache on the runners.

I ended up implementing a lock at the API level to prevent another parallel chains from being added as a temporary solution (chains are added async). Potentially there is a way around this in re-designing the system to be more scalable. The simple way I was hoping for was if celery could queue chains after one another.
